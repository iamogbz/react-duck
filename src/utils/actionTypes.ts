const randomString = () => Math.random().toString(36).substring(7);
const ns = (s: TemplateStringsArray) => `@@context/${s}/${randomString()}`;

export const ActionTypes = {
    INIT: ns`INIT`,
    PROBE_UNKNOWN_ACTION: ns`PROBE_UNKNOWN_ACTION`,
};
