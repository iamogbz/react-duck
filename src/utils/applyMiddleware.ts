// github.com/reduxjs/redux/blob/8551ba8/src/applyMiddleware.ts
import { compose } from "./compose";

export function applyMiddleware<S, T extends string, P>(
    ...middlewares: Middleware<S, T, P>[]
): ContextEnhance<S, T, P> {
    return function enhancer(
        context: ContextValue<S, T, P>,
    ): ContextValue<S, T, P> {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        function dispatchStub(...args: unknown[]): never {
            throw new Error(
                "Dispatching while constructing your middleware is not allowed. " +
                    "Other middleware would not be applied to this dispatch.",
            );
        }

        const middlewareAPI: MiddlewareAPI<S, T, P> = {
            dispatch: dispatchStub,
            getState: context.getState,
        };
        const chain = middlewares.map((middleware) =>
            middleware(middlewareAPI),
        );
        const dispatch = compose<typeof context.dispatch>(...chain)(
            context.dispatch,
        );
        return {
            ...context,
            dispatch,
        };
    };
}
