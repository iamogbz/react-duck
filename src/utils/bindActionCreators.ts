// github.com/reduxjs/redux/blob/208d7f1/src/bindActionCreators.ts

export function bindActionCreator<T extends string, P>(
    actionCreator: ActionCreator<T, P>,
    dispatch: ContextDispatch<T, P>,
): ActionDispatcher<P> {
    return function dispatchAction(...args: P[]): void {
        dispatch(actionCreator(...args));
    };
}

/**
 * Turns an object whose values are action creators, into an object with the
 * same keys, but with every function wrapped into a `dispatch` call so they
 * may be invoked directly.
 */
export function bindActionCreators<T extends string, P>(
    actionCreator: ActionCreator<T, P>,
    dispatch: ContextDispatch<T, P>,
): ActionDispatcher<P>;

export function bindActionCreators<T extends string, P>(
    actionCreator: ActionCreator<T, P>,
    dispatch: ContextDispatch<T, P>,
): ActionDispatcher<P>;

export function bindActionCreators<T extends string, P, S>(
    actionCreators: ActionCreatorMapping<T, P, S>,
    dispatch: ContextDispatch<T, P>,
): ActionDispatcherMapping<T, P>;

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export function bindActionCreators(
    actionCreators: ActionCreator | ActionCreatorMapping,
    dispatch: ContextDispatch,
) {
    if (typeof actionCreators === "function") {
        return bindActionCreator(actionCreators, dispatch);
    }

    if (typeof actionCreators !== "object" || actionCreators === null) {
        throw new Error(
            `bindActionCreators expected an object or a function, instead received ${
                actionCreators === null ? "null" : typeof actionCreators
            }. ` +
                'Did you write "import ActionCreators from" instead of "import * as ActionCreators from"?',
        );
    }

    const boundActionCreators: ActionDispatcherMapping = {};
    for (const key in actionCreators) {
        const actionCreator = actionCreators[key];
        if (typeof actionCreator === "function") {
            boundActionCreators[key] = bindActionCreator(
                actionCreator,
                dispatch,
            );
        }
    }
    return boundActionCreators;
}
