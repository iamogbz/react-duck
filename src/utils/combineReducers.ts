type State = any; // eslint-disable-line @typescript-eslint/no-explicit-any

export function combineReducers<State>(
    initialState: Record<string, State>,
    reducerMapping: ReducerMapping<State>,
): Reducer<Record<string, State>> {
    const reducers = Object.entries(reducerMapping);
    return function (
        state: Record<string, State> = initialState,
        action,
    ): Record<string, State> {
        return reducers.reduce((acc, [name, reducer]) => {
            acc[name] = reducer(state[name], action);
            return acc;
        }, {} as Record<string, State>);
    };
}
