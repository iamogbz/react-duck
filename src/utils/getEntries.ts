export function getEntries<A extends string, B>(o: Record<A, B>): [A, B][] {
    return Object.entries(o) as [A, B][];
}
