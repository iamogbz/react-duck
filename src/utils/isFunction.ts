export function isFunction<F>(maybeFunction: F | unknown): maybeFunction is F {
    return typeof maybeFunction === "function";
}
