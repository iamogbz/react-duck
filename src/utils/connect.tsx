import * as React from "react";
import { GlobalContext } from "src/components/Context";
import { bindActionCreators } from "./bindActionCreators";

function defaultMergeProps<T extends string, P, I, K, J>(
    stateProps?: K,
    dispatchProps?: ActionDispatcherMapping<T, P>,
    ownProps?: I,
): J {
    return ({ ...ownProps, ...stateProps, ...dispatchProps } as unknown) as J;
}

function isFunction<F>(maybeFunction: F | unknown): maybeFunction is F {
    return typeof maybeFunction === "function";
}

function asMapDispatchToPropsFn<S, T extends string, P, I>(
    actionCreators?: ActionCreatorMapping<T, P, S>,
): MapDispatchToProps<T, P, I> | undefined {
    return (
        actionCreators &&
        function mapDispatchToProps(
            dispatch: ContextDispatch<T, P>,
        ): ActionDispatcherMapping<T, P> {
            return bindActionCreators(actionCreators, dispatch);
        }
    );
}

export function connect<
    S,
    I,
    T extends string,
    P,
    K,
    J extends I & K & ActionDispatcherMapping<T, P>
>(
    mapStateToProps?: MapStateToProps<S, I, K>,
    mapDispatchToProps?:
        | MapDispatchToProps<T, P, I, ActionDispatcherMapping<T, P>>
        | ActionCreatorMapping<T, P, S>,
    mergeProps?: (
        stateProps?: K,
        dispatchProps?: ActionDispatcherMapping<T, P>,
        ownProps?: I,
    ) => J,
    options?: ConnectOptions<S, T, P, I, K, J>,
): (component: ReactComponent<J>) => React.FunctionComponent<I> {
    const mapDispatchToPropsFn = isFunction<MapDispatchToProps<T, P, I>>(
        mapDispatchToProps,
    )
        ? mapDispatchToProps
        : asMapDispatchToPropsFn(mapDispatchToProps);

    return function wrapWithConnect(
        componentToWrap: ReactComponent<J>,
    ): React.FunctionComponent<I> {
        const WrappedComponent = (options?.pure
            ? React.memo(componentToWrap)
            : componentToWrap) as ReactComponent<J>;
        const wrappedComponentName =
            componentToWrap.displayName || componentToWrap.name || "Component";
        const displayName = `Connect(${wrappedComponentName})`;

        function WrapperComponent(
            props: I,
        ): React.ReactElement<J, typeof WrappedComponent> {
            const { state, dispatch } = React.useContext(
                options?.context ?? GlobalContext,
            );
            const stateProps = React.useMemo(
                () => mapStateToProps?.(state, props),
                [props, state],
            );
            const dispatchProps = React.useMemo(
                () => mapDispatchToPropsFn?.(dispatch, props),
                [dispatch, props],
            );
            const mergedProps = React.useMemo(
                () =>
                    (mergeProps ?? defaultMergeProps)(
                        stateProps,
                        dispatchProps,
                        props,
                    ),
                [dispatchProps, props, stateProps],
            );
            return <WrappedComponent {...mergedProps} />;
        }

        WrapperComponent.WrappedComponent = WrappedComponent;
        WrapperComponent.displayName = displayName;
        return WrapperComponent;
    };
}
