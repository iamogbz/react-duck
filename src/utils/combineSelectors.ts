import { getEntries } from "./getEntries";

export function combineSelectors<
    S,
    T extends string,
    P,
    R,
    N extends string,
    Q extends string
>(
    duckName: N,
    selectors?: SelectorMapping<S, R, T, P, Q>,
): Nullable<SelectorMapping<Record<N, S>, R, T, P, Q>> {
    if (!selectors) return;
    const duckSelectors = {} as SelectorMapping<Record<N, S>, R, T, P, Q>;
    for (const [q, s] of getEntries(selectors)) {
        duckSelectors[q] = (state) => s(state[duckName]);
    }
    return duckSelectors;
}
