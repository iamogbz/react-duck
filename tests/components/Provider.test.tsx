import * as React from "react";
import { RenderResult, act, cleanup, render } from "@testing-library/react";
import {
    GlobalContext,
    Provider,
    applyMiddleware,
    createContext,
    createDuck,
    createRootDuck,
    createRootProvider,
    useDispatch,
    useSelector,
} from "src";
import { ActionTypes } from "src/utils/actionTypes";

describe("provider", () => {
    const DECREMENT = "decrement";
    const INCREMENT = "increment";
    const decrement = (state: number) => state - 1;
    const increment = jest.fn((state: number) => state + 1);
    const counterReducer: (s: number, a?: Action<string>) => number = (
        state,
        action,
    ) => {
        switch (action?.type) {
            case DECREMENT:
                return decrement(state);
            case INCREMENT:
                return increment(state);
            default:
                return state;
        }
    };
    const counterDuck = createDuck({
        initialState: 0,
        name: "counter",
        reducers: { [DECREMENT]: counterReducer, [INCREMENT]: counterReducer },
        selectors: { get: (s) => s["counter"] },
    });

    const ACTION_TYPE_INIT = "init";
    const init = jest.fn(() => true);
    const initDuck = createDuck({
        actionMapping: { [ActionTypes.INIT]: ACTION_TYPE_INIT },
        initialState: false,
        name: "init",
        reducers: { [ACTION_TYPE_INIT]: init },
    });

    const dummyDuck = createDuck({
        initialState: null,
        name: "dummy",
        reducers: {},
    });

    const rootDuck = createRootDuck(counterDuck, initDuck, dummyDuck);

    const dummyMiddleware: Middleware<
        typeof rootDuck.initialState,
        ContextDispatch<Action>
    > = () => (next) => (action) => next(action);

    const Context = createContext(
        rootDuck.reducer,
        rootDuck.initialState,
        applyMiddleware(dummyMiddleware),
        "GlobalContext",
        true,
    );

    const RootProvider = createRootProvider(Context);

    function Example() {
        const increment = useDispatch(counterDuck.actions.increment);
        const init = useSelector(rootDuck.selectors.init!.$, Context);
        const count = useSelector(counterDuck.selectors?.get);
        return (
            <div>
                Count: <span>{count}</span>
                <button disabled={!init} onClick={increment}>
                    increment
                </button>
            </div>
        );
    }

    afterEach(() => {
        increment.mockClear();
        init.mockClear();
        jest.clearAllMocks();
        cleanup();
    });

    describe("context", () => {
        it("renders without root provider", async () => {
            const result = render(<Example />);
            // eslint-disable-next-line jest/prefer-snapshot-hint
            expect(result.baseElement).toMatchSnapshot();
            const element = await result.findByText("increment");
            act(() => element.click());
            expect(increment).not.toHaveBeenCalled();
        });

        it("renders with root provider and updates on action dispatch", async () => {
            const result = render(
                <RootProvider>
                    <Example />
                </RootProvider>,
            );
            const element = await result.findByText("increment");
            act(() => element.click());
            expect(increment).toHaveBeenCalledWith(0);
            act(() => element.click());
            expect(increment).toHaveBeenCalledWith(1);
            // eslint-disable-next-line jest/prefer-snapshot-hint
            expect(result.baseElement).toMatchSnapshot();
            expect(init).toHaveBeenCalledTimes(1);
        });

        it("renders with enhanced context", async () => {
            const loggerMiddleware: Middleware<
                typeof rootDuck.initialState,
                ContextDispatch<Action>
            > =
                ({ getState }) =>
                (next) =>
                async (action) => {
                    // eslint-disable-next-line no-console
                    console.log("action to dispatch", action);
                    // Call the next dispatch method in the middleware chain.
                    const returnValue = await next(action);
                    // eslint-disable-next-line no-console
                    console.log("state after dispatch", getState());
                    // This will likely be the action itself, unless
                    // a middleware further in chain changed it.
                    return returnValue;
                };
            const EnhancedContext = createContext(
                rootDuck.reducer,
                rootDuck.initialState,
                applyMiddleware(dummyMiddleware, loggerMiddleware),
                "EnhancedContext",
            );

            const spyConsoleLog = jest
                .spyOn(console, "log")
                .mockImplementation(() => undefined);
            render(
                <Provider Context={EnhancedContext}>
                    <Example />
                </Provider>,
            );
            await Promise.resolve();
            expect(spyConsoleLog).toHaveBeenCalledTimes(2);
            expect(spyConsoleLog.mock.calls[0]).toMatchObject([
                "action to dispatch",
                {
                    payload: undefined,
                    type: expect.stringContaining("@@context/INIT"),
                },
            ]);
            expect(spyConsoleLog.mock.calls[1]).toMatchObject([
                "state after dispatch",
                { counter: 0, init: true },
            ]);
            spyConsoleLog.mockRestore();
        });

        it("fails to render if middleware dispatches while constructing", () => {
            const badMiddleware: Middleware<
                typeof rootDuck.initialState,
                ContextDispatch<Action>
            > = ({ dispatch }) => {
                dispatch({ type: "SOME_ACTION" });
                return () => async (action) => action;
            };
            const emptyRootDuck = createRootDuck();
            const ErrorContext = createContext(
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                emptyRootDuck.reducer as Reducer<any, Action>,
                emptyRootDuck.initialState,
                applyMiddleware(dummyMiddleware, badMiddleware),
                "ErrorContext",
            );
            const spyConsoleError = jest
                .spyOn(console, "error")
                .mockImplementation(() => undefined);
            expect(() =>
                render(
                    <Provider Context={ErrorContext}>
                        <Example />
                    </Provider>,
                ),
            ).toThrow(
                "Dispatching while constructing your middleware is not allowed",
            );
            spyConsoleError.mockRestore();
        });
    });

    describe("observable", () => {
        async function runAssertions(
            result: RenderResult,
            listener: jest.Mock,
        ) {
            expect(listener).toHaveBeenCalledTimes(2);
            const button = await result.findByText("increment");
            act(() => button.click());
            expect(listener).toHaveBeenCalledTimes(3);
            act(() => button.click());
            expect(listener).toHaveBeenCalledTimes(4);
            expect(listener).toHaveBeenLastCalledWith(
                expect.objectContaining({
                    dispatch: expect.any(Function),
                    getState: expect.any(Function),
                    reducer: expect.any(Function),
                    state: {
                        counter: 2,
                        dummy: null,
                        init: true,
                    },
                    subscribe: expect.any(Function),
                }),
            );
            const [[lastCallArg0]] = listener.mock.calls.slice(-1);
            expect(lastCallArg0[Symbol.observable]()).toBe(lastCallArg0);
        }

        it("subscribes successfully to context value changes from component", async () => {
            expect.assertions(5);
            const listener = jest.fn();
            function Sample() {
                const value = React.useContext(GlobalContext);
                React.useEffect(
                    function contextSubscribe() {
                        value.subscribe(listener);
                    },
                    [value],
                );
                const increment = useDispatch(
                    rootDuck.actions.counter.increment,
                );
                const init = useSelector(rootDuck.selectors.init!.$);
                return (
                    <div>
                        <button disabled={!init} onClick={increment}>
                            increment
                        </button>
                    </div>
                );
            }
            const result = render(
                <RootProvider>
                    <Sample />
                </RootProvider>,
            );
            await runAssertions(result, listener);
        });

        it("subscribes successfully to context value changes from enhancer", async () => {
            expect.assertions(6);
            const listener = jest.fn();
            const supplement = { extra: "value" };
            const enhancer = jest.fn(function enhancer(value) {
                value.subscribe(listener);
                return { ...value, ...supplement };
            });
            const Context = createContext(
                rootDuck.reducer,
                rootDuck.initialState,
                enhancer,
            );
            function Sample() {
                const increment = useDispatch(
                    rootDuck.actions.counter.increment,
                    Context,
                );
                const init = useSelector(rootDuck.selectors.init!.$, Context);
                return (
                    <div>
                        <button disabled={!init} onClick={increment}>
                            increment
                        </button>
                    </div>
                );
            }
            const result = render(
                <Provider Context={Context}>
                    <Sample />
                </Provider>,
            );
            await runAssertions(result, listener);
            expect(listener).toHaveBeenLastCalledWith(
                expect.objectContaining(supplement),
            );
        });
    });
});
