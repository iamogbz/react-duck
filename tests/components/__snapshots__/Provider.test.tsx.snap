// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider context renders with root provider and updates on action dispatch 1`] = `
<body>
  <div>
    <div>
      Count: 
      <span>
        2
      </span>
      <button>
        increment
      </button>
    </div>
  </div>
</body>
`;

exports[`provider context renders without root provider 1`] = `
<body>
  <div>
    <div>
      Count: 
      <span>
        0
      </span>
      <button
        disabled=""
      >
        increment
      </button>
    </div>
  </div>
</body>
`;
